### PR TITLE
Installing profiles to the compliance-service

### PR DESCRIPTION
**Pull request by lancewf**
### :nut_and_bolt: Description: What code changed, and why?
Finishing the functionality of when a user clicks the "install" button on a profile content item. When this happens the profile is installed for the current login in user only. This add the compliance-service as a dependency on the automate-cds service. 

### :chains: Related Resources

https://github.com/chef/automate/issues/3900

### :athletic_shoe: How to Build and Test the Change
1. `build components/automate-deployment && build components/automate-gateway && build components/automate-cds && build components/automate-ui-devproxy && start_all_services && start_automate_ui_background && ui_logs`
1. After the UI finishes starting log in as admin and go to the profiles page (https://a2-dev.test/compliance/compliance-profiles).
1. Ensure there are no profiles install. 
1. Go to the CDS page (https://a2-dev.test/cds). 
1. Click "Install" on one or more profile content items. 
1. Ensure the success banner is displayed. 
1. Go back to the profiles page (https://a2-dev.test/compliance/compliance-profiles) and ensure those profiles are installed. 

### :white_check_mark: Checklist

- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [ ] Tests added/updated?
- [ ] Docs added/updated?
- [x] All commits have been signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: api/external/cds/request/root.proto [hunk 1]
```diff
@@ -9,6 +9,7 @@ message ContentItems {
 
 message InstallContentItem {
   string id = 1;
+  string request_user = 2;
```
Comment (lancewf) [inline api/external/cds/request/root.proto:12]: The user the profile is installed for.

File: components/automate-cds/backend/client.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
+package backend
```
Comment (lancewf) [inline components/automate-cds/backend/client.go:1]: This is a mock backend that provides the fake content items

File: components/automate-cds/config/config.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+	LogFormat         string `mapstructure:"log-format"`
+	LogLevel          string `mapstructure:"log-level"`
+	certs.TLSConfig   `mapstructure:"tls"`
+	ComplianceService `mapstructure:"compliance-service"`
+}
+
+// ComplianceService connection information
+type ComplianceService struct {
+	Address string `mapstructure:"address"`
+}
```
Comment (lancewf) [inline components/automate-cds/config/config.go:17]: Adding the compliance-service

File: components/automate-cds/habitat/config/config.toml [hunk 4]
```diff
@@ -6,3 +6,9 @@ log-level = "{{cfg.log.level}}"
 cert_path = "{{pkg.svc_config_path}}/service.crt"
 key_path = "{{pkg.svc_config_path}}/service.key"
 root_ca_path ="{{pkg.svc_config_path}}/root_ca.crt"
+
+
+[compliance-service]
+{{~#eachAlive bind.compliance-service.members as |member|}}
+address = "{{member.sys.ip}}:{{member.cfg.port}}"
+{{~/eachAlive}}
```
Comment (lancewf) [inline components/automate-cds/habitat/config/config.toml:14]: Adding the compliance-service

File: components/automate-cds/habitat/plan.sh [hunk 5]
```diff
@@ -14,6 +14,9 @@ pkg_exports=(
   [port]=service.port # default service is grpc
   [host]=service.host
 )
+pkg_binds=(
+  [compliance-service]="port"
+)
 pkg_exposes=(port)
 
 pkg_bin_dirs=(bin)
```
Comment (lancewf) [inline components/automate-cds/habitat/plan.sh:22]: Adding the compliance-service

File: components/automate-cds/habitat/plan.sh [hunk 6]
```diff
@@ -14,6 +14,9 @@ pkg_exports=(
   [port]=service.port # default service is grpc
   [host]=service.host
 )
+pkg_binds=(
+  [compliance-service]="port"
+)
```
Comment (lancewf) [inline components/automate-cds/habitat/plan.sh:19]: Adding the compliance-service

File: components/automate-cds/server/root.go [hunk 7]
```diff
@@ -46,63 +59,87 @@ func (s *Server) GetVersion(
...
+
+		err = s.installProfile(ctx, contentItem, request.RequestUser)
+		if err != nil {
+			return &response.InstallContentItem{}, status.Errorf(codes.Internal,
+				"Error Installing Profile: %s", err.Error)
+		}
+	} else {
+		return &response.InstallContentItem{}, status.Errorf(codes.InvalidArgument,
+			"Can install content item type %q", contentItem.Type)
+	}
```
Comment (lancewf) [inline components/automate-cds/server/root.go:122]: Filled out the install function more to provide errors for content items that can not be installed.

File: components/automate-cds/server/root.go [hunk 8]
```diff
@@ -121,3 +158,44 @@ func (s *Server) DownloadContentItem(request *request.DownloadContentItem,
...
+
+	_, err = stream.CloseAndRecv()
+	if err != nil {
+		return status.Errorf(codes.Internal,
+			"Failed closing stream to the compliance service error: %q", err.Error())
+	}
+
+	return nil
+}
```
Comment (lancewf) [inline components/automate-cds/server/root.go:201]: This is actually sending the profile to the compliance-service.

File: components/automate-cds/server/server.go [hunk 9]
```diff
@@ -7,29 +7,45 @@ import (
...
+		// This should never happen
+		log.WithError(err).Error("Failed to create compliance-service connection")
+		return err
+	}
+	defer complianceConn.Close() // nolint: errcheck
+
+	profilesServiceClient := profiles.NewProfilesServiceClient(complianceConn)
+
+	serv := NewGRPCServer(s, profilesServiceClient).Serve(list)
+	s.Logger.Debugf("Automate CDS GRPC API listening on %v", config.GRPC)
```
Comment (lancewf) [inline components/automate-cds/server/server.go:49]: Setting up the connection to the compliance-service and passing that connection to the cds server.

File: components/automate-deployment/pkg/assets/binds.go [hunk 10]
```diff
@@ -13,6 +13,7 @@ automate-builder-api BINDING_MODE strict
 automate-builder-api-proxy REQUIRED automate-builder-api
 automate-builder-api-proxy BINDING_MODE strict
 automate-builder-memcached BINDING_MODE strict
+automate-cds REQUIRED compliance-service
```
Comment (lancewf) [inline components/automate-deployment/pkg/assets/binds.go:16]: It is so easy to add a service to another service now. All I had to do is add the binding to the plan.sh file and run `deployment_generate`. 

Good job deployment team!

**Comment by vjeffrey** [comment]
hmmm, i keep hitting the following error...
```
Error: Failed to start service: Failed to load service vj/automate-cds/0.1.0/20200702222929
Output:
✗✗✗
✗✗✗ [Err: 7] Missing required bind(s), compliance-service
✗✗✗
: exit status 1

Deploy failed
DeploymentServiceCallError: A request to the deployment-service failed: Request to stream deployment events failed: deploy failed
```

**Comment by lancewf** [comment]
@vjeffrey Sorry I forgot to list that the automate-deployment needs build also. I will update it now.

**Comment by vjeffrey** [comment]
👍

**Pull request merged**